In [288]:
text = "Король японских забегаловок, прародитель доширака, та самая лапша из «Наруто» и кучи других тайтлов. Также помимо этого есть легенда, что рамен был ввезен китайским учёным Чжу Сунсуи, который был вынужден стать беженцем и служил как советник Токугаве Мицукуни, японскому феодальному лорду. Часто используется темпура удон и тофу (кицунэ удон). Вдохновением для блюда была другая закуска — «акашияки», которая готовилась в префектуре Хъёго, городе Акаши, и была сделана из яиц и осьминога. Первые такояки содержали в себе конняку и говядину, однако потом Эндо переключился снова на осьминога. В начале 20 века карри появилось в рационе японского морского флота, как способ борьбы с авитаминозом, и также используется по сей день (в одном аниме есть серия, посвящённая карри, как раз потому, что там корабледевочки). Вообще есть небольшой лайфхак. Нужно купить обычное порошковое карри, тут важно почитать состав, карри бывают разные, вам нужен тот, где есть фенугрек или фенхель или оба сразу. Был я значит в этой вашей Японии, честно пытался схарчевать там рамен. В обоих случаях оглядываясь по сторонам я заметил что самураи уплетали это явство с явным удовольствием и причмокиванием. Ну вот приготовление якитори они познали в совершенстве, с этим не поспоришь. Ну и для взрослых конечно это японский вискарь, офигительно крутая штука. Сакэ норм но на любителя, правда получше корейской сётю. И все равно не могу понять в чем разница между карри и тушенной картохой, кроме специи. Готовил и карри, и якисобу, и разные вариации рамена, омурайсу, онигири, кацудон. Нынешняя ДОМАШНЯЯ японская кухня довольно простая и понятная, всё делается из очень доступных продуктов. Ресторанка и фастфуд - уже совсем другая песня."

Данный текст подходит для оценки, поскольку в нем есть  слова с «неизвестными» корнями, не содержащие продуктивных аффиксов, для которых можно однозначно определить лемму и часть
речи. Также здесь есть несколько имен собственных и неологизмов (схарчевать, корабледевочки). Несколько жаргонов также присутствуют (картоха, вискарь, норм, ресторанка). 

In [289]:
import csv
import nltk
from nltk.tokenize import word_tokenize

nltk.download('punkt')
from string import punctuation
words = [w.lower() for w in word_tokenize(text) if w.isalpha()]
#print(words)
with open("words.csv", "w") as file:
  for elem in words:
    elem_new = elem.split()
    writer = csv.writer(file)
    writer.writerow(elem_new)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


Для разметки я беру тэгсет как в pymorphy. Это из-за того, что он более полно может разделить части речи. Но в этом есть и минус. Когда я приводила все к одному виду, пришлось использовать тэги из других тэгсетов (udpipe), поскольку многие тэггеры работают именно на этой коллекции. Поэтому я решила разметить корпус еще и по стандарту universal dependencies. 


Здесь корпус, размеченный по типа **pymorphy** и конвертированный

In [347]:
my_parse = []
with open ('words_tags.csv') as f:
  reader = csv.reader(f, delimiter = ';')
  for row in reader:
    my_parse.append((row[0], convert(row[1])))
  my_parse.remove(my_parse[0])
  print(my_parse)

[('король', 'NOUN'), ('японских', 'ADJ'), ('забегаловок', 'NOUN'), ('прародитель', 'NOUN'), ('доширака', 'NOUN'), ('та', 'ADJ'), ('самая', 'ADJ'), ('лапша', 'NOUN'), ('из', 'ADP'), ('наруто', 'NOUN'), ('и', 'CCONJ'), ('кучи', 'NOUN'), ('других', 'ADJ'), ('тайтлов', 'NOUN'), ('также', 'CCONJ'), ('помимо', 'ADP'), ('этого', 'ADJ'), ('есть', 'VERB'), ('легенда', 'NOUN'), ('что', 'CCONJ'), ('рамен', 'NOUN'), ('был', 'VERB'), ('ввезен', 'PRTS'), ('китайским', 'ADJ'), ('учёным', 'NOUN'), ('чжу', 'NOUN'), ('сунсуи', 'NOUN'), ('который', 'ADJ'), ('был', 'VERB'), ('вынужден', 'PRTS'), ('стать', 'VERB'), ('беженцем', 'NOUN'), ('и', 'CCONJ'), ('служил', 'VERB'), ('как', 'CCONJ'), ('советник', 'NOUN'), ('токугаве', 'NOUN'), ('мицукуни', 'NOUN'), ('японскому', 'ADJ'), ('феодальному', 'ADJ'), ('лорду', 'NOUN'), ('часто', 'ADV'), ('используется', 'VERB'), ('темпура', 'NOUN'), ('удон', 'NOUN'), ('и', 'CCONJ'), ('тофу', 'NOUN'), ('кицунэ', 'ADJ'), ('удон', 'NOUN'), ('вдохновением', 'NOUN'), ('для', 'AD

Здесь корпус, размеченный по типу **universal dependencies**

In [348]:
my_parse2 = []
with open ('pos_tags_ud.csv') as f:
  reader = csv.reader(f, delimiter = ';')
  for row in reader:
    my_parse2.append((row[0], convert(row[1])))
  my_parse2.remove(my_parse2[0])
  print(my_parse2)

[('король', 'NOUN'), ('японских', 'ADJ'), ('забегаловок', 'NOUN'), ('прародитель', 'NOUN'), ('доширака', 'NOUN'), ('та', 'DET'), ('самая', 'ADJ'), ('лапша', 'NOUN'), ('из', 'ADP'), ('наруто', 'NOUN'), ('и', 'CCONJ'), ('кучи', 'NOUN'), ('других', 'ADJ'), ('тайтлов', 'NOUN'), ('также', 'CCONJ'), ('помимо', 'ADP'), ('этого', 'DET'), ('есть', 'AUX'), ('легенда', 'NOUN'), ('что', 'SCONJ'), ('рамен', 'NOUN'), ('был', 'AUX'), ('ввезен', 'VERB'), ('китайским', 'ADJ'), ('учёным', 'NOUN'), ('чжу', 'NOUN'), ('сунсуи', 'NOUN'), ('который', 'ADJ'), ('был', 'AUX'), ('вынужден', 'VERB'), ('стать', 'VERB'), ('беженцем', 'NOUN'), ('и', 'CCONJ'), ('служил', 'VERB'), ('как', 'SCONJ'), ('советник', 'NOUN'), ('токугаве', 'NOUN'), ('мицукуни', 'NOUN'), ('японскому', 'ADJ'), ('феодальному', 'ADJ'), ('лорду', 'NOUN'), ('часто', 'ADV'), ('используется', 'VERB'), ('темпура', 'NOUN'), ('удон', 'NOUN'), ('и', 'CCONJ'), ('тофу', 'NOUN'), ('кицунэ', 'ADJ'), ('удон', 'NOUN'), ('вдохновением', 'NOUN'), ('для', 'ADP')

Создаем список кортежей для всех теггеров

**PyMorphy**

In [291]:
from pymorphy2 import MorphAnalyzer
morph = MorphAnalyzer()

In [ ]:
'''
unique_words = []
for word in words:
  if word not in unique_words:
    unique_words.append(word)
len(unique_words)
'''

'\nunique_words = []\nfor word in words:\n  if word not in unique_words:\n    unique_words.append(word)\nlen(unique_words)\n'

In [292]:
def pymorphy_parse(words_list):
  new_list = []
  for word in words_list:
    p = morph.parse(word)[0]
    tag = p.tag.POS
    new_list.append((word, convert(tag)))
  return new_list

**Natasha**


In [293]:
from natasha import NewsMorphTagger, NewsEmbedding, Doc, Segmenter
emb = NewsEmbedding()
morph_tagger = NewsMorphTagger(emb)
segmenter = Segmenter()

In [294]:
def natasha_parse(word_list):
  new_list = []
  new_text = ' '.join(word_list)
  doc = Doc(new_text)
  doc.segment(segmenter)
  doc.tag_morph(morph_tagger)
  #print(doc.tokens)
  for elem in doc.tokens:
    #print(elem.pos)
    tag = elem.pos
    word = elem.text
    new_list.append((word, tag))
  return new_list

**Stanza**

In [ ]:
import stanza
stanza.download('ru')


2021-10-03 15:26:37 INFO: Downloading default packages for language: ru (Russian)...
2021-10-03 15:26:39 INFO: File exists: /root/stanza_resources/ru/default.zip.
2021-10-03 15:26:46 INFO: Finished downloading models and saved to /root/stanza_resources.


In [295]:
nlp = stanza.Pipeline(lang='ru', processors='pos, tokenize')
def stanza_parse(word_list):
  new_list = []
  new_text = ' '.join(word_list)  
  doc = nlp(new_text)
  for sent in doc.sentences:
    for word in sent.words:
      new_list.append((word.text,word.upos))
  return new_list
print(stanza_parse(words))

2021-10-03 16:52:52 INFO: Loading these models for language: ru (Russian):
| Processor | Package   |
-------------------------
| tokenize  | syntagrus |
| pos       | syntagrus |

2021-10-03 16:52:52 INFO: Use device: cpu
2021-10-03 16:52:52 INFO: Loading: tokenize
2021-10-03 16:52:52 INFO: Loading: pos
2021-10-03 16:52:53 INFO: Done loading processors!


[('король', 'NOUN'), ('японских', 'ADJ'), ('забегаловок', 'NOUN'), ('прародитель', 'NOUN'), ('доширака', 'NOUN'), ('та', 'DET'), ('самая', 'ADJ'), ('лапша', 'NOUN'), ('из', 'ADP'), ('наруто', 'NOUN'), ('и', 'CCONJ'), ('кучи', 'NOUN'), ('других', 'ADJ'), ('тайтлов', 'NOUN'), ('также', 'ADV'), ('помимо', 'ADP'), ('этого', 'PRON'), ('есть', 'VERB'), ('легенда', 'NOUN'), ('что', 'SCONJ'), ('рамен', 'NOUN'), ('был', 'AUX'), ('ввезен', 'VERB'), ('китайским', 'ADJ'), ('учёным', 'NOUN'), ('чжу', 'NOUN'), ('сунсуи', 'PUNCT'), ('который', 'PRON'), ('был', 'AUX'), ('вынужден', 'VERB'), ('стать', 'VERB'), ('беженцем', 'NOUN'), ('и', 'CCONJ'), ('служил', 'VERB'), ('как', 'SCONJ'), ('советник', 'NOUN'), ('токугаве', 'NOUN'), ('мицукуни', 'NOUN'), ('японскому', 'ADJ'), ('феодальному', 'ADJ'), ('лорду', 'NOUN'), ('часто', 'ADV'), ('используется', 'VERB'), ('темпура', 'NOUN'), ('удон', 'NOUN'), ('и', 'CCONJ'), ('тофу', 'NOUN'), ('кицунэ', 'NOUN'), ('удон', 'NOUN'), ('вдохновением', 'NOUN'), ('для', 'AD

Для приведения всех тегов к общему виду я сделала 4 множества из уникальных тегов для каждого размеченного корпуса и при составлении списка кортежей использовала функцию конвертации.

In [300]:
def get_unique_pos(word_list):
  elems = []
  for elem in word_list:
    if elem == None:
      continue
    elems.append(elem[1])
  un_set = set(elems)
  return un_set

In [350]:
natasha_pos = get_unique_pos(natasha_tags)
stanza_pos = get_unique_pos(stanza_tags)
pymorphy_pos = get_unique_pos(pymorphy_tags)
my_pos = get_unique_pos(my_parse)
#print(natasha_pos) 
#print(stanza_pos)
#print( pymorphy_pos)
#print( my_pos)
#print((pymorphy_pos)-(natasha_pos|stanza_pos))

In [309]:
def convert(word):
  if word == None:
    return 'None'
  w = word.upper()
  dict_uns = { 'NPRO': 'PRON', 
   'CONJ' : 'CCONJ' ,
   'PRCL' : 'PART',
  'ADJS' : 'ADJ',
   'INFN' : 'VERB',
   'NUMR' : 'NUM',
   'ADVB' : 'ADV',
   'PREP' : 'ADP',
   'ADJF' : 'ADJ'
  }

  if word in dict_uns.keys():
    return dict_uns[w]
  return w

Далее для каждого получившегося списка считаем accuracy

In [279]:
!pip install sklearn

In [280]:
from sklearn.metrics import accuracy_score

In [281]:
def get_lists(list_texts):
  return [elem[1] for elem in list_texts]

In [284]:
def get_accuracy(expected_list, actual_list):
  return accuracy_score(expected_list, actual_list)

Здесь **accuracy**, посчитанный для корпуса, размеченного для тэгсета по типу **pymorphy**. Как видим, сам Pymorphy проанализировал лучше всех, при этом Stanza не сильно ему уступает

In [312]:
pymorphy_tags = pymorphy_parse(words)
natasha_tags = natasha_parse(words)
stanza_tags = stanza_parse(words)
my_tags = get_lists(my_parse)
nat_my = get_accuracy(my_tags, get_lists(natasha_tags))
print("Natasha: ", nat_my)
py_my = get_accuracy(my_tags, get_lists(pymorphy_tags))
print("PyMorphy: ", py_my)
stan_my = get_accuracy(my_tags, get_lists(stanza_tags))
print("Stanza: ", stan_my)

Natasha:  0.7431906614785992
PyMorphy:  0.8715953307392996
Stanza:  0.8054474708171206


Здесь **accuracy**, посчитанный для корпуса, размеченного по тэгсету **universal dependencies**. Stanza здесь отработал лучше всего.

In [346]:
my_tags2 = get_lists(my_parse2)
nat_my2 = get_accuracy(my_tags2, get_lists(natasha_tags))
print("Natasha: ", nat_my2)
py_my2 = get_accuracy(my_tags2, get_lists(pymorphy_tags))
print("PyMorphy: ", py_my2)
stan_my2 = get_accuracy(my_tags2, get_lists(stanza_tags))
print("Stanza: ", stan_my2)

Natasha:  0.7821011673151751
PyMorphy:  0.7937743190661478
Stanza:  0.8560311284046692


Возьмем сначала несколько типов n-грамм для pymorphy, а затем несколько для stanza

In [351]:
def chunker(words_list, tags_list):
  n = len(tags_list)
  res = []
  for i in range(len(words_list)-n+1):
    suc = True
    for j in range(n):    
      if words_list[i+j][1] != tags_list[j]:
          suc = False
          break
    if suc:
      res.append([elem[0] for elem in words_list[i:i+n]])
  return res
print("Pymorphy: ")
print(chunker(pymorphy_tags, ['NOUN', "ADV"]) )
print(chunker(pymorphy_tags, ['NOUN', "ADV", "VERB"]))
print(chunker(pymorphy_tags, ['ADJ', "NOUN"]))
print("Stanza: ")
print(chunker(stanza_tags, ['NOUN', "ADV"]) )
print(chunker(stanza_tags, ['NOUN', "ADV", "VERB"]))
print(chunker(stanza_tags, ['ADJ', "NOUN"]))

Pymorphy: 
[['лорду', 'часто'], ['раз', 'потому'], ['корабледевочки', 'вообще'], ['карри', 'тут'], ['японии', 'честно'], ['кухня', 'довольно'], ['фастфуд', 'уже']]
[['лорду', 'часто', 'используется'], ['корабледевочки', 'вообще', 'есть'], ['японии', 'честно', 'пытался']]
[['японских', 'забегаловок'], ['самая', 'лапша'], ['других', 'тайтлов'], ['китайским', 'учёным'], ['феодальному', 'лорду'], ['другая', 'закуска'], ['первые', 'такояки'], ['морского', 'флота'], ['сей', 'день'], ['одном', 'аниме'], ['небольшой', 'лайфхак'], ['порошковое', 'карри'], ['вашей', 'японии'], ['явным', 'удовольствием'], ['крутая', 'штука'], ['корейской', 'сётю'], ['разные', 'вариации'], ['японская', 'кухня'], ['доступных', 'продуктов'], ['другая', 'песня']]
Stanza: 
[['тайтлов', 'также'], ['лорду', 'часто'], ['говядину', 'однако'], ['раз', 'потому'], ['корабледевочки', 'вообще'], ['карри', 'тут'], ['японии', 'честно'], ['взрослых', 'конечно'], ['вискарь', 'офигительно'], ['любителя', 'правда'], ['кухня', 'довол

Про прошлую домашку:

Думаю, что качество улучшили бы шаблоны:

*   "ADV" + "NOUN" -  очень хорошая
*   "ADJ" + "NOUN" -  хорошая маска
*   "не" + "VERB" -  не понравилась

Люди часто употребляют слова очень хорошая/плохая/красивая и т.д., когда пишут отзывы, особенно в сочетании с наречием. Скорее это будет чаще встречаться у положительных отзывов, поскольку к плохи вещам зачастую люди оставляют более эксперссивные отзывы (ужасный, отвратительный). Второй шаблон нужен для того, чтобы определить значение прилагательного. Так, например, в положительных отзывах мы скорее встретим отличная маска, приятный аромат, а в отрицательных ужасный запах. Третий шаблон будет нам помогать для выявления хороших и плохих словосочетаний со словом не. Обычно люди пишут, что им что-то не нравится, но иногда не вносит наоборот положительную окраску - не прогадала.



Ниже код из прошлой домашки + немного добавление от меня для поиска биграмм по определенным шаблонам (простите за не лучший код)

In [352]:
import operator, csv
import random
import nltk
from nltk.tokenize import word_tokenize
nltk.download("punkt")
nltk.download('stopwords')
from nltk.corpus import stopwords
from pymorphy2 import MorphAnalyzer
morph = MorphAnalyzer()
def get_sorted():
  with open('revs.csv') as csvfile:
    reader = csv.reader(csvfile, delimiter=';')
    for row in reader:
      sortedlist = sorted(reader, key=operator.itemgetter(0), reverse=True)
  return sortedlist


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [353]:

def norm_token_review(review):
  result = []
  words = [w.lower() for w in word_tokenize(review) if w.isalpha()]
  for word in words:
      result.append(morph.parse(word)[0].normal_form) 
  return result

def tokenize_normalize(list_revs):
  trueforms = []
  for item in list_revs:
    words = norm_token_review(item[1])
    trueforms.append([item[0], words])
  return trueforms

In [354]:
def almost_true():
  truelist=tokenize_normalize(get_sorted())
  tryy = []
  almost_good = truelist[:40]
  reversee = truelist[::-1]
  alm_bad = reversee[:40]
  return almost_good, alm_bad

In [355]:
v1, v2 = almost_true()

In [ ]:
v2

In [383]:
def get_str(w_list):
  words_list = []
  words2 = ''
  for elem in w_list:
    words2 += ' '.join(elem[1])+ " "
  return words2


In [390]:
def chunker_words(revss):
  res = []
  for i in range(len(revss)-1):    
    if revss[i][0] == "не" and revss[i+1][1] == "VERB":
      res.append(revss[i][0]+" "+revss[i+1][0])
  return res

создаю списки по шаблонам

In [411]:
revs_parse_plus = stanza_parse([get_str(v1)])
revs_parse_minus = stanza_parse([get_str(v2)])

pos_a_n = chunker(revs_parse_plus, ['ADV', "NOUN"])
pos_adj_n = chunker(revs_parse_plus, ['ADJ', "NOUN"])
pos_ne = chunker_words(revs_parse_plus)

neg_a_n = chunker(revs_parse_minus, ['ADV', "NOUN"])
neg_adj_n = chunker(revs_parse_minus, ['ADJ', "NOUN"])
neg_ne = chunker_words(revs_parse_minus)
print(neg_ne)

['не получить', 'не заметить', 'не быть', 'не купить', 'не тратить', 'не заметить', 'не очищать', 'не вытягивать', 'не понравиться', 'не купить', 'не вытягивать', 'не разглаживать', 'не выдержать', 'не быть', 'не делать', 'не стоить', 'не отставать', 'не выйти', 'не советовать', 'не советовать', 'не любить', 'не очистить', 'не стать', 'не советовать', 'не понравиться', 'не рекомендовать', 'не ложиться', 'не делать', 'не рекомендовать', 'не понравиться', 'не понравиться', 'не совпадать', 'не подойти', 'не хотеть', 'не знать', 'не подойти', 'не стать', 'не иметь', 'не увидеть', 'не быть', 'не говорить', 'не заметить', 'не взять', 'не заметить', 'не купить', 'не заметить', 'не знать', 'не увидеть', 'не понравиться', 'не поравиться', 'не измениться', 'не ошибаться', 'не отваливаться', 'не сказать', 'не мочь', 'не заметить', 'не иметь', 'не пахнуть']


Делаю из списков множество

In [445]:
def get_sets(set1,set2, set3):
  bigs_n = set1+set2
  bigs_all_neg = []
  for elem in bigs_n:
    if type(elem) == list:
      new_elem = ' '.join(elem)
      bigs_all_neg.append(new_elem)
  neg_set = set(bigs_all_neg + set3)
  return neg_set
pos_set = get_sets(pos_a_n, pos_adj_n, pos_ne)
neg_set = get_sets(neg_a_n, neg_adj_n, neg_ne)

Здесь я получила два множества - хороших биграмм и плохих

In [446]:
good_set = pos_set - neg_set
bad_set = neg_set - pos_set
good_set, bad_set

({'альгинатный маска',
  'большой счёт',
  'включая лицо',
  'внешний сторона',
  'воздушный текстура',
  'гладкий кожа',
  'густой сыворотка',
  'ещё раз',
  'жёлтый кувшинка',
  'кдивительный морщина',
  'конечно эффект',
  'корейский косметология',
  'крутой маска',
  'лекарственный гриб',
  'лёгкий массаж',
  'много пропитка',
  'много сыворотка',
  'много эссенуий',
  'много эссенция',
  'не вызвать',
  'не давать',
  'не жечь',
  'не оставлять',
  'не разгладить',
  'не раздражать',
  'не смывать',
  'не сползать',
  'не стекать',
  'не сфотографировать',
  'не щипит',
  'нежный аромат',
  'отличный маска',
  'отличный серия',
  'офигенный ощущение',
  'первый раз',
  'положительный эмоция',
  'прикольный маска',
  'приятно центелла',
  'приятный маска',
  'приятный масочок',
  'проблемный кожа',
  'прохладноц вода',
  'рабочий неделя',
  'роскошный эффект',
  'склизкий консистенция',
  'специальный средство',
  'столовый ложка',
  'суперскай маска',
  'сухой последствие',
  'тка

Узнаем, хороший отзыв или плохой. True - отзыв положительный (оценка 5 или 4), False - отрицательный (все остальные оценки). А также подсчет качества модели

In [447]:
def is_good_review(review, good_set, bad_set):
  norm_rev = norm_token_review(review)
  revs_parse = stanza_parse(norm_rev)
  a_n = chunker(revs_parse, ['ADV', "NOUN"])
  adj_n = chunker(revs_parse, ['ADJ', "NOUN"])
  ne = chunker_words(revs_parse)
  n_set = get_sets(a_n, adj_n, ne)
  #print(n_set)
  is_good = n_set & good_set
  is_bad = n_set & bad_set
  if len(is_good) >= len(is_bad):
    return True
  else:
    return False

In [448]:
y_pred = []
y_true = []
reviews = [['5', 'Хороший уход. Хорошо увлажняет и питает кожу. Работает ночью, пользуюсь маской почти месяц.'], ['4', 'Маска с пудрой винограда и розы, и кембрийской глиной. Просто добавь воды. :)'], ['3', 'Странная маска, которая делает кожу лица только хуже, но прекрасно работает в качестве скраба для тела. Неожиданный поворот!'], ['2', 'Маска из разряда "ни о чем". Бессмысленный продукт.'], ['2', 'Бесполезный продукт с сильным запахом. Думаю, не для сухой кожи его делали. Скорее для жирной подойдет, чтоб подсушить ее.'], ['5', 'Приятная маска, не обещает волшебства и полностью справляется с заявленными функциями.'], ['5', 'Очень-очень приятное знакомство!) Средства с классным запах и к тому же реально работающие.'], ['3', 'Маска оказалась для меня практически бесполезной.'], ['1', 'Маска, когда наносишь на лицо жжёт, ужасный запах, как будто покрышка горит, плохой состав, снимается, грубо говоря с кожей.'], ['5', 'Отличные маски!В упаковке было 40 штук - 10 штук было в подарок! Эффект мне нравится - успокаивает кожу, снимает раздражения и красноту. Использую вечером на ночь, липкости нет, утром кожа приятная, мягкая.Японское качество, что тут скажешь']]
for rnd in reviews:
  review = rnd[1]
  score = rnd[0]
  y_true.append(int(score) > 3)
  y_pred.append(is_good_review(review, good_set, bad_set))
print(y_true)
print(y_pred)
accuracy_score(y_true, y_pred)

[True, True, False, False, False, True, True, False, False, True]
[True, True, True, True, True, True, True, True, False, True]


0.6

при внедрении в программу биграмм accuracy стала ниже. До этого было 0.8, сейчас стала 0.6. Вполне возиожно, что это из-за того, что тест проводился на маленьком объеме данных. Попробуем провести на тех же данных, что и в прошлой домашке


In [449]:
y_pred1 = []
y_true1 = []
reviews = get_sorted()
for i in range (100):
  rnd = random.choice(reviews)
  review = rnd[1]
  score = rnd[0]
  y_true1.append(int(score) > 3)
  y_pred1.append(is_good_review(review, good_set, bad_set))
print(y_true1)
print(y_pred1)
accuracy_score(y_true1, y_pred1)

[True, False, False, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, False, True, True, True, False, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, False, False, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True]
[True, False, False, True, True, True, True, True, True, True, True, True, True, True, True, False, True, False, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, False, True, False, True, True, True, True, True, True, False, True, True, False, False, False, True, True, True, False, 

0.85

Accuracy стал выше - с 0.8 до 0.85, на некоторых данных доходил до 0.86